enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,761 @@
+	// '*' can be used at the beginning of the wildcard DNS name. For example, '*.example.com.'
+	// will match 'sub1.example.com.' but won't match 'sub2.sub1.example.com.'
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	Name string `json:"name"`
+}
+
+// DNSNameResolverStatus defines the observed status of DNSNameResolver.
+type DNSNameResolverStatus struct {
+	// resolvedNames contains a list of matching DNS names and their corresponding IP addresses
+	// along with TTL and last DNS lookup time.
+	// +listType=map
+	// +listMapKey=dnsName
+	// +patchMergeKey=dnsName
+	// +patchStrategy=merge
+	// +optional
+	ResolvedNames []DNSNameResolverStatusItem `json:"resolvedNames,omitempty" patchStrategy:"merge" patchMergeKey:"dnsName"`
+}
+
+// DNSNameResolverStatusItem describes the details of a resolved DNS name.
+type DNSNameResolverStatusItem struct {
+	// dnsName is the resolved DNS name matching the name field of DNSNameResolverSpec.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	DNSName string `json:"dnsName"`
+	// ips contains the list of IP addresses associated with the dnsName.
+	// +kubebuilder:validation:Required
+	// +listType=set
+	IPs []string `json:"ips"`
+	// ttlSeconds is the minimum time-to-live value among all the IP addresses.
+	// +kubebuilder:validation:Required
+	TTLSeconds int32 `json:"ttlSeconds"`
+	// lastLookupTime is the timestamp when the last DNS lookup was completed.
+	// +kubebuilder:validation:Required
+	LastLookupTime *metav1.Time `json:"lastLookupTime"`
+	// resolutionFailures keeps the count of how many times the DNS resolution failed for the
+	// dnsName field. If the DNS resolution succeeds then the field will be set to zero. Upon
+	// every failure, the value of the field will be incremented by one. Upon reaching a threshold
+	// value, the details about the DNS name will be removed.
+	ResolutionFailures int `json:"resolutionFailures,omitempty"`
+	// conditions provide information about the state of the DNS name.
+	//
+	// These are the supported conditions:
+	//
+	//   * Degraded
+	//   - True if the following conditions are met:
+	//     * The last DNS name resolution failed.
+	//   - False if any of those conditions are unsatisfied.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+````
+
+### Workflow Description
+
+The workflows for the different components related to DNSNameResolver events are explained in this section. The workflow
+for those events is shown in the following diagram with an example `DNSNameResolver` CR:
+
+![Workflow](./coredns-egressfirewall-integration.png)
+
+#### Workflow of OVN-K cluster manager
+
+The following pseudocode explains the workflow of the OVN-K cluster manager:
+```shell
+watch(EgressFirewall)
+
+if match(event.type, "create"):
```

Comment:
Typically Kube controllers don't distinguish between create/update/delete operations and instead just look at the state of the world for any event and reconcile the difference. You aren't guaranteed to see create/delete events so I would expect your update logic to have to handle those cases anyway.

Is the logic planned to be implemented as separate event handlers or will there be reconciliation logic as described above?

Reply:
OVN-K has a different retry framework which does distinguish between the different events and also retries actions on objects. I am planning to use the same that's why separated them out: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/retry/obj_retry.go